### ISSUE TITLE
Test-Json  string format not matching

### ISSUE THREAD

**Issue by FormindAPE** [PowerShell/PowerShell#19888]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Let's create a very simple json schema that use the string format `duration` :
```json
{
	"type": "object",
	"additionalProperties":false,
	"properties": {
		"duration": {
			"type": "string",
			"format": "duration"
		}
	}
}
```

the format `duration` was added in the json schema "draft 2019-09" current version is "draft 2020-12". The goal of the `duration` format is to match ISO 8601 (see: [json schema documentation](https://json-schema.org/understanding-json-schema/reference/string.html#format) )

and save it in a file named `schema.json`

and run the following command to test:
```powershell
> Test-Json -SchemaFile .\schema.json -Json '{"duration":"test123"}'
```
or
```powershell
> Test-Json -SchemaFile .\schema.json -Json '{"duration":"PT1D"}'
```

the result will be `True` when expecting `False`

### Expected behavior

```console
PS> Test-Json -SchemaFile .\schema.json -Json '{"duration":"test123"}'
False
PS> Test-Json -SchemaFile .\schema.json -Json '{"duration":"PT1D"}'
False


PS> Test-Json -SchemaFile .\schema.json -Json '{"duration":"P1D"}'
True
```

### Actual behavior

```console
PS> Test-Json -SchemaFile .\schema.json -Json '{"duration":"test123"}'
True
PS> Test-Json -SchemaFile .\schema.json -Json '{"duration":"PT1D"}'
True


PS> Test-Json -SchemaFile .\schema.json -Json '{"duration":"P1D"}'
True
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0-preview.4
PSEdition                      Core
GitCommitId                    7.4.0-preview.4
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by brian6932** [comment]
There seems to be an issue regarding sending `"` within strings to programs that are not native pwsh, they aren't getting escaped in `7.4.0-preview.4`

**Comment by FormindAPE** [comment]
the same behavior happens when piping a file to the command:
```powershell
> Get-content .\schema.json
{
        "type": "object",
        "additionalProperties":false,
        "properties": {
                "duration": {
                        "type": "string",
                        "format": "date"
                }
        }
}
> Get-content .\data.json
{"duration":"test123"}
> Get-content .\data.json | Test-Json -SchemaFile .\schema.json
True
```

**Comment by brian6932** [comment]
Can you please revise the title of this issue, as it's a bit too specific, it'll likely cause many copies of it

**Comment by FormindAPE** [comment]
> There seems to be an issue regarding sending `"` within strings to programs that are not native pwsh, they aren't getting escaped in `7.4.0-preview.4`, no matter the string type you use `""...""`and `'"..."'` will result in the same thing

@brian6932 
I got 2 different behavior when using `'"..."'` and `""...""` (see below) so I think it's not related 
```powershell
PS > echo "{"duration":"test"}" | Test-Json -SchemaFile .\schema.json        
Test-Json: Cannot parse the JSON.
False
Test-Json: Cannot parse the JSON.
False
PS > echo '{"duration":"test"}' | Test-Json -SchemaFile .\schema.json        
True
```

**Comment by FormindAPE** [comment]
After some more digging I have noticed that none of the json shema string format options are working in  7.4.0-preview.4
but that in 7.3.5 moste of the string format option were working (only some  option like `duration`, `regex` were not woking)

**Comment by gregsdennis** [comment]
In JSON Schema, `format ` is an annotation by default.  That means it doesn't validate.

The two solutions I see here are:

1. Just internally configure JsonSchema.Net to validate `format` by default.  This may align with expected behavior, but you wouldn't be able to turn it off.
2. Add a new PowerShell parameter that enables that configuration, e.g. `-ValidateFormat`.

The fix is pretty easy; we just need to decide which to do.

**Comment by iSazonov** [comment]
@gregsdennis Thanks! I hope WG will do a conclusion. /cc @SteveL-MSFT

**Comment by brian6932** [comment]
`7.4.0-preview.6` fixes the issue I had

**Comment by gregsdennis** [comment]
Uh... what was the fix? Where was it implemented (e.g. a PR)?

**Comment by brian6932** [comment]
Not sure, but I don't have any issues with nesting quote types anymore '{"...":"..."}' seems to parse just find now 🤷

**Comment by gregsdennis** [comment]
Oh, so your problem is with quotes in general, which isn't what this issue is about.

**Comment by brian6932** [comment]
It seemed like the same underlying cause

**Comment by TobiasPSP** [comment]
Is there still the need for a fix/WG discussion, or was is just a quoting issue?

**Comment by pearj** [comment]
> Is there still the need for a fix/WG discussion, or was is just a quoting issue?

@TobiasPSP This is still definitely a problem, and the quotes issue @brian6932 was experiencing was a red herring I believe.

> After some more digging I have noticed that none of the json shema string format options are working in 7.4.0-preview.4
but that in 7.3.5 moste of the string format options were working (only some option like duration or regex were not woking)

I agree with @FormindAPE that it was mostly working in 7.3.x, I tried 7.3.6 that I had lying around and `date-time` was working but `duration` was not.

Using this `schema.json`:

```json
{
  "type": "object",
  "additionalProperties": false,
  "properties": {
    "duration": {
      "type": "string",
      "format": "duration"
    },
    "datetest": {
      "type": "string",
      "format": "date-time"
    }
  }
}
```

In PowerShell 7.3.6:

```powershell
PS > Test-Json -SchemaFile .\schema.json -Json '{"datetest":"test123"}'
Test-Json: DateTimeExpected: #/datetest
False

> Test-Json -SchemaFile .\schema.json -Json '{"duration":"test123"}' 
True
```

In PowerShell 7.4.0:

```powershell
PS > Test-Json -SchemaFile .\schema.json -Json '{"datetest":"test123"}'
True

> Test-Json -SchemaFile .\schema.json -Json '{"duration":"test123"}' 
True
```

> The two solutions I see here are:
> 
> 1. Just internally [configure](https://docs.json-everything.net/schema/basics/#schema-format) JsonSchema.Net to validate `format` by default.  This may align with expected behavior, but you wouldn't be able to turn it off.
> 2. Add a new PowerShell parameter that enables that configuration, e.g. `-ValidateFormat`.
> 
> The fix is pretty easy; we just need to decide which to do.

@gregsdennis My vote is for doing 1, in a patch version of PowerShell 7.4.x, since this is essentially a possible unintentional breaking change from 7.3.x. then do 2 in maybe 7.5 or something like that?

**Comment by gregsdennis** [comment]
I don't mind doing both of these as a single update, but I would need someone from the PS team to confirm.

**Comment by gregsdennis** [comment]
cc: @iSazonov ☝️

**Comment by iSazonov** [comment]
From document referenced above:
> A JSON Schema validator will ignore any format type that it does not understand.

So I think it is optional feature for the cmdlet too. In any case we need WG conclusion.

**Comment by gregsdennis** [comment]
> From document referenced above:
> 
> > A JSON Schema validator will ignore any format type that it does not understand. - @iSazonov 

To be clear, the spec requires that `format` is an annotation unless the implementation is configured to validate or the `format-assertion` vocabulary is specified by the meta-schema.

Since, generally, the default meta-schemas do not reference the `format-assertion` vocab, it falls to the configuration of the lib.

@pearj is arguing that previous versions of PowerShell validated `format`, implying that such configuration was enabled by default, and that `format` now not being validated is a breaking change which should be reverted (enable `format` validation by default).

I'm happy to have to enabled by default, if that's what _PowerShell_ wants to do, but it needs to have a switch to disable it.  (Personally I feel this is backwards, but I understand the historical context.)

**Comment by pearj** [comment]
> Personally I feel this is backwards, but I understand the historical context

@gregsdennis Yes you are right, going to the WG sounded like it could take a long time, so I was hoping to restore functionality in a quicker timeframe.  But if either change requires WG approval then off by default probably makes the most sense then.

**Comment by TobiasPSP** [comment]
Cmdlet WG has reviewed this. In #20782 we recommended adding the parameter `-Option` to `Test-Json` to allow for more optional leniency. `-Option` takes a *string array*. We feel that `Test-Json` should by default perform strict validation according to the Json Schema. The ability to *optionally* include `Format` in the validation process is an added benefit that should be exposed through an additional option item in  `-Option`.
